### ISSUE TITLE
Allow a timezone configuration for partial date handling

### ISSUE THREAD

**Issue by mshick** [contentlayerdev/contentlayer#9]
Currently, if I set a field to the `date` type and enter a partial date, e.g., `2021-10-21` I'm given a date object initialized with a value in UTC, like this `2021-10-21T00:00:00.000Z`. Ideally I'd be able to have that partial date interpreted correctly for my timezone — I'm in `America/New_York` so I'd like to see `2021-10-21T04:00:00.000Z`.

A possible, straightforward solution would be the `date-fns-tz` libraries `zonedTimeToUtc` function and an optional IANA timezone in the contentlayer config. E.g.

```js
zonedTimeToUtc(doc.createdAt, opts.timezone)
```

If you're interested in a contribution I could work on a PR.

**Comment by schickling** [comment]
Hey @mshick I've just released a new `dev` version that introduces a `date.timezone` configuration option.

Please give it a try via `yarn add contentlayer@dev next-contentlayer@dev` and:

```ts
export default makeSource({
  contentDirPath: 'content',
  documentTypes: [Doc],
  date: { timezone: 'America/New_York' },
})
```

I'll release a stable version with this feature soon.

**Comment by mshick** [comment]
Hi @schickling very sorry for the silence, haven't been able to touch my contentlayer project over the last week.

Thanks so much for the change! I'll test out today and add my results to this issue. I'll go ahead an close since it looks like you've got this.

**Comment by mshick** [comment]
Apologies for the long delay in testing this, but I was trying out the feature today and noticed a couple things:

**Minor issue**

The `makeSource` fn is being exported without the correct types for this feature. It looks like the `PartialArgs` type in `core` was updated, but `makeSource` is exported with a custom `Args` type passed to the generic.

**Larger issue**

The utility of this option doesn't work as I would have imagined. In my custom solution for this issue I treated my field `publishedAt` as a string, and passed that string to `zonedTimeToUtc`. This has the result of interpreting the string value in the specific timezone, and then doing the timezone offset. So, the partial date string is converted from `2020-10-20` to `2021-10-20T04:00:00.000Z`.  This is due to the line this line: https://github.com/contentlayerdev/contentlayer/blob/589aa04099f6d096064cb699b5af95a771e75e05/packages/%40contentlayer/source-files/src/fetchData/mapping.ts#L244

If that date object were created only if we don't need to use `zonedTimeToUtc` then things should work correctly assuming I specify the frontmatter field in my markdown as an explicit string.

```
  publishedAt: '2020-10-20'
```

If I add the field without the quotes though, I think `gray-matter` is turning it into a `Date` object before you touch it. I'm not sure if it's even an option, but since you apply typing explicitly in `contentlayer` it might make sense to have gray-matter treat the field as a string no matter what, so that you can control when / how it is turned into a `Date`, and can offer a consistent user experience for the new `timezone` option — meaning, they can get the TZ applied whether they specify the partial date as a quoted string or an unquoted date-string.

**Comment by pard68** [comment]
Hate to neco-bump, but I think I have found a possible solution. Like you had mentioned @mshick , the issue is upstream of Contentlayer. I did some digging and it's actually one layer above gray-matter, since gray-matter doesn't actually parse anything at all, it just gets the right stuff extracted to be parsed by another library (JSON, YAML, etc).

And that's the problem -- at least for YAML (but I presume others) the definition of the YAML standard states that a date string is implicitly UTC. This means there isn't really any upstream changes that we can really effect since we'd be changing the YAML standard!

So my potential solution is that if the `options.date.timezone` value exists we do some work to transpose the time. There doesn't seem to be a drop in solution from another library (like `date-fns`) to do what we're looking for, but I think we can pull it off simply enough to not have a need for another library anyway.

Instead of trying to convert our already UTC date to UTC (this won't work) like here:
https://github.com/contentlayerdev/contentlayer/blob/589aa04099f6d096064cb699b5af95a771e75e05/packages/%40contentlayer/source-files/src/fetchData/mapping.ts#L244

Instead we need to transpose the date as such:
```
              if (options.date?.timezone) {
                dateValue = new Date(dateValue.getTime() + dateFnsTz.getTimezoneOffset(options.date.timezone) * -1)
              }
```

Basically we take the existing UTC date, get the epoch time from it, and then _add_ the given timezone's offset from UTC in seconds. We have to "flip" the offset though because we're assuming the original date is in the requested timezone already and instead we want to turn it into a UTC timestamp.

I created a PR that does essentially this, albeit with a slightly less naive approach.

**Comment by schickling** [comment]
This should be addressed with the [`0.2.7` release](https://github.com/contentlayerdev/contentlayer/releases/tag/v0.2.7). 🎉
